lib/spack/spack/package.py

```diff
@@ -719,20 +719,38 @@ def version_urls(self):
         return version_urls
 
     def nearest_url(self, version):
-        """Finds the URL for the next lowest version with a URL.
-           If there is no lower version with a URL, uses the
-           package url property. If that isn't there, uses a
-           *higher* URL, and if that isn't there raises an error.
+        """Returns the nearest URL for a target version.
+
+        The algorithm to compute the URL is the following:
+
+            1. If the target version is associated with a specific URL, that
+            URL will be returned.
+
+            2. Otherwise return the URL associated with the closest higher
+            version.
+
+            3. If there's no such a version return the default URL from
+            the package.
+
+        The idea of the algorithm is that the package-level URL should be
+        the default one for recent versions, and that once a version
+        overrides its URL that override is valid for all the versions that
+        are older.
+
+        Args:
+            version (Version): the target version
+
+        Returns:
+            Nearest URL
         """
-        version_urls = self.version_urls()
-        url = getattr(self.__class__, 'url', None)
-
-        for v in version_urls:
-            if v > version and url:
-                break
-            if version_urls[v]:
-                url = version_urls[v]
-        return url
+        candidates = sorted([x for x in self.version_urls() if x >= version])
+
+        # If we don't have candidates, return the default. Here we let
```

Comment:
maybe catch the error and emit a custom message? Not sure if this should be done here or in `package.py` though. But the end results ideally would be that the error message contains the version and package the url should have been computed for

Reply:
Hmm, I would be more for maybe documenting that this function might raise an `AttributeError`. If need be we could catch it in `url_for_version`. Waiting for @tgamblin and @adamjstewart  opinion on that.

Reply:
I would make this consistent with the existing behavior of returning `None` and look into making the errors more consistent in another PR.   There are places where Spack checks whether it can get a `url` from a package by looking for `None` and I don't remember if raising an error here will break anything.